keys, sql: change sequence key format so backup/restore works

vilterp opened the pull request:
Before, a sequence's key was of the format `/Table/55/seqval`. Now, it's `/Table/55/0/0/0`. This mimics the structure of a table key, allowing backup/restore to work without special casing for sequences.

Keeping the original key format would have required special casing sequences in the restore implementation's keyrewriter code, which is already complicated; this approach seems to minimize complexity.

Fixes: #21354

Release note: None

Commit:
keys, sql: change sequence key format so backup/restore works

Before, a sequence's key was of the format /Table/55/seqval. Now, it's
/Table/55/0/0/0. This mimics the structure of a table key, allowing
backup/restore to work without special casing for sequences.

Keeping the original key format would have required special casing
sequences in the restore implementation's keyrewriter code, which is
already complicated; this approach seems to minimize complexity.

Release note (enterprise change, backward-incompatible change):
Enable backup/restore for sequences. This changes how sequences
are stored in the KV layer, so existing sequences will need to be
dropped and recreated. Since a sequence cannot be dropped while it
is being used in a column's DEFAULT expression, those expressions
will need to be dropped before the sequence is dropped, and recreated
after the sequence is recreated. The `setval` function can be used to
set the value of a sequence to what it was previously.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21684)
<!-- Reviewable:end -->

mjibson commented:
Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup_test.go, line 2537 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3W4sajtl4N-Kf1nSFp:-L3W4sajtl4N-Kf1nSFq:bx4lure) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/pkg/ccl/sqlccl/backup_test.go#L2537)):*
> ```Go
> 	sqlDB.Exec(t, `INSERT INTO data.t (v) VALUES ('foo')`)
> 	sqlDB.Exec(t, `INSERT INTO data.t (v) VALUES ('bar')`)
> 	sqlDB.Exec(t, `INSERT INTO data.t (v) VALUES ('baz')`)
> ```

Use sqlDB.QueryStr here to save the results of a SELECT...

---

*[pkg/ccl/sqlccl/backup_test.go, line 2545 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3W4eF9OkkCb_tCMwnY:-L3W4eF9OkkCb_tCMwnZ:b4dqoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/pkg/ccl/sqlccl/backup_test.go#L2545)):*
> ```Go
> 	// Verify that the db was restored correctly.
> 	{
> 		expectedRows := []struct {
> ```

Then use sqlDB.CheckQueryResults to verify they were restored correctly. This is much easier than all the stuff you have here.

---

*[pkg/sql/create_sequence.go, line 142 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3W5G7QXHak5hngFDYk:-L3W5G7QXHak5hngFDYl:b-cnemms) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/pkg/sql/create_sequence.go#L142)):*
> ```Go
> 		{
> 			ID:   0,
> 			Name: "primary",
> ```

sqlbase.PrimaryKeyIndexName

---

*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 420 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3W5Td06TOvhayABP5A:-L3W5Td06TOvhayABP5B:bg7z1sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/pkg/sql/logictest/testdata/logic_test/pg_catalog#L420)):*
> ```
> ----
> primary         i
> primary         i
> ```

Is this really correct?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684)*
<!-- Sent from Reviewable.io -->

benesch commented:
Sequences have been appearing in release notes, right? This seems like it should be listed as a backwards-incompatible change.

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/20171102_sql_sequences.md, line 204 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3WE_4XbhWLXzgprlJi:-L3WE_4XbhWLXzgprlJj:btwo6lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/docs/RFCS/20171102_sql_sequences.md#L204)):*
> ```Markdown
> Where the zeroes are dummy values for the index ID, primary key value, and
> column family ID, respectively. This mimics the structure of a normal table
> key, so backup/restore can work without modification.
> ```

Ooh, this should go in the encoding tech note when you have a minute.

---

*[pkg/ccl/sqlccl/backup_test.go, line 2537 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3W4sajtl4N-Kf1nSFp:-L3WEPXtgKuoKVINrG4l:bknjjqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/pkg/ccl/sqlccl/backup_test.go#L2537)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Use sqlDB.QueryStr here to save the results of a SELECT...
</blockquote></details>

Psst, SQL supports multiple rows per INSERT:

```
INSERT INTO data.t (v) VALUES ('foo'), ('bar'), ('baz');
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-:-L3WEhD5S5Ua96BcpQ5S:bz69mjt)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Yeah, we have to at least call this out as a backwards-incompatible change. If someone was running the previous alpha, created a sequence, and then upgraded to this version, what happens? Does it transparently reset to zero or are they left with a broken sequence? (can they even drop it and recreate it to recover?) 

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-:-L3YiVsGJjjmIZmBjsMX:b-qd0yy9)*
<!-- Sent from Reviewable.io -->

vilterp commented:
It transparently resets to zero. They can then set it to what it should be with `SELECT setval('my_seq', (SELECT max(my_seq_col) + 1 FROM my_tbl))`. If you do a live upgrade with an INSERT workload on a table which uses a sequence for primary keys, you'll have constraint violation errors as you try to insert and use values that have already been used. So you'll have to take it down during the upgrade. Hopefully this won't be a problem for alpha users.

Will call this out & explain behavior in the release notes. @bdarnell do you think this requires writing a migration?

bdarnell commented:
Will the orphaned key cause any problems (for example, will it prevent you from backing up or restoring in the future)? I think this is probably OK without a migration.

vilterp commented:
@bdarnell Yes, the old key would stick around and would cause restores to crash, since they wouldn't know how to rewrite it.

On second thought, I decided that keeping the old format (`/Table/<id>/"seqval"`) might be worth teaching the key rewriter how to rewrite that format. The old format makes it more clear that it's a sequence. Other things like SCRUB which assume that something in the table keyspace will have a table-like key structure (indexes, etc) will break on sequences even with the new key format, since the dummy indexes and column families don't actually exist, so might as well be explicit.

New approach (teach key rewriter about sequence keys) here: #21766

mjibson commented:
Does SCRUB still need to be taught about "/seqval" is a key if #21766 is merged? A previous version of this PR had a function that would create some fake indexes on the descriptor. Can that solution be adapted to appease SCRUB?

vilterp commented:
Testing scrub now; let me see…

Commit:
sql: add logic test for scrub on sequences

`SCRUB TABLE seq` errors because it asserts that it's scrubbing an
actual table, not a sequence. But, `SCRUB DATABASE` goes over the
sequence and seems perfectly happy.

Release note: None

vilterp commented:
Ok, bouncing back to this approach…

Added the fake index back in (this time properly as the primary index), plus a fake column, and a fake family. Calling `showCreateTable` on a sequence descriptor actually returns a well-formed `CREATE TABLE` statement at this point, though that's not tested anywhere (what is tested is that `crdb_internal.create_statements` contains a well-formed `CREATE SEQUENCE` statement for the sequence).

This brings us pretty close to a sequence being a one-row, one-column table, except:
- The value isn't SQL-encoded (just a bare int) — I don't know the details of SQL encoding here
- You can't INSERT, UDPATE, DELETE it
- You can't DROP TABLE or ALTER TABLE it (have to use DROP SEQUENCE and ALTER SEQUENCE)

mjibson commented:
If we do decide to change the key structure to be encoded like other keys, we should ensure that a sequence can never appear to be a table to a user. That is, nothing in crdb_internal and no `SHOW CREATE TABLE` statement should work on it, even if it does act and look like a table to the system.

vilterp commented:
Currently:
- [x] `SHOW CREATE TABLE my_seq` errors out
- [x] `SHOW CREATE SEQUENCE` works. (https://github.com/cockroachdb/cockroach/blob/92f352913fcdabda3dc115f56da8481a24af30fb/pkg/sql/logictest/testdata/logic_test/sequences#L112-L116)
- sequences show up in `crdb_internal.tables`. I guess this is bad? Again, awkward — it's sort of a table.
- sequences show up in `crdb_internal.create_statements`, but with "sequence" in the `descriptor_type` column (this is sometimes also "view").

Commit:
sql: assert that SHOW CREATE TABLE doesn't work on a sequence

Release note: None

Commit:
ccl: handle when only one of a (table, sequence) pair is restored

A table using a sequence in the DEFAULT expression of one of its columns
creates a dependency from that table to that sequence. Handle restoring
just one or the other in the same way that foreign key references are
already handled in RESTORE.

Release note: None

vilterp commented:
- Added to the encoding tech note
- Added handling for when only one of a (table, sequence) pair is backed up
- Addressed other comments

PTAL! Thanks.

---

Review status: 1 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/RFCS/20171102_sql_sequences.md, line 204 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3WE_4XbhWLXzgprlJi:-L3plO8aBOy6esEQEbs0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/docs/RFCS/20171102_sql_sequences.md#L204)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ooh, this should go in the encoding tech note when you have a minute.
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup_test.go, line 2537 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3W4sajtl4N-Kf1nSFp:-L3plQg-RXHlQIwaa-80:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/pkg/ccl/sqlccl/backup_test.go#L2537)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Psst, SQL supports multiple rows per INSERT:

```
INSERT INTO data.t (v) VALUES ('foo'), ('bar'), ('baz');
```
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup_test.go, line 2545 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3W4eF9OkkCb_tCMwnY:-L3plRINUzSSKRoHry-I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/pkg/ccl/sqlccl/backup_test.go#L2545)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Then use sqlDB.CheckQueryResults to verify they were restored correctly. This is much easier than all the stuff you have here.
</blockquote></details>

Done.

---

*[pkg/sql/create_sequence.go, line 142 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3W5G7QXHak5hngFDYk:-L3plSb2VHklqWmBrUD8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/pkg/sql/create_sequence.go#L142)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

sqlbase.PrimaryKeyIndexName
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 420 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L3W5Td06TOvhayABP5A:-L3pnIY2MgGFIPtoxWZY:b-xfrznq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61c5b72057524083223401aad97fb721fa3c312/pkg/sql/logictest/testdata/logic_test/pg_catalog#L420)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Is this really correct?
</blockquote></details>

No, This index doesn't show up in PG output. Prevented it from showing up here with a change to `pg_catalog.go`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-:-L3pnQZm_p2sMZsTIMq9:b8crypo)*
<!-- Sent from Reviewable.io -->

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 8 files at r1, 8 of 11 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: 11 of 13 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup_test.go, line 2605 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L4CyZdx3TZlgy_Jrqi5:-L4CyZdx3TZlgy_Jrqi6:bkt5pqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b990298b42126d16202e64051698878fe833c82c/pkg/ccl/sqlccl/backup_test.go#L2605)):*
> ```Go
> }
> 
> func TestBackupRestoreSequence(t *testing.T) {
> ```

revision_history is the one thing missing here, but i'm fine adding that as part of my general addition of tests on revision_history in #21717 

---

*[pkg/keys/keys.go, line 689 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L4Csx6HzaogL56duXbp:-L4Csx6HzaogL56duXbq:b9lmcj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b990298b42126d16202e64051698878fe833c82c/pkg/keys/keys.go#L689)):*
> ```Go
> func MakeSequenceKey(tableID uint32) []byte {
> 	key := MakeTablePrefix(tableID)
> 	key = encoding.EncodeUvarintAscending(key, SequenceIndexID)        // Index id
> ```

this is still a constant suffix right? just more readable to reconstruct it inline?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-:-L4CvSxacEUeErnW7v4D:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vilterp commented:
TFTRs!

---

Review status: 11 of 13 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup_test.go, line 2605 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L4CyZdx3TZlgy_Jrqi5:-L4D1Z6W_hceC8WEKuDA:b9ldnc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b990298b42126d16202e64051698878fe833c82c/pkg/ccl/sqlccl/backup_test.go#L2605)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

revision_history is the one thing missing here, but i'm fine adding that as part of my general addition of tests on revision_history in #21717 
</blockquote></details>

Cool, thanks.

---

*[pkg/keys/keys.go, line 689 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-L4Csx6HzaogL56duXbp:-L4D2bsbc3U5rIKCNoU8:b-824q9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/b990298b42126d16202e64051698878fe833c82c/pkg/keys/keys.go#L689)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

this is still a constant suffix right? just more readable to reconstruct it inline?
</blockquote></details>

Putting a `SequenceSuffix` in constants.go might be good, but I think keeping it here is readable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21684#-:-L4D2heBr9SyRUd-5pVJ:bz5mair)*
<!-- Sent from Reviewable.io -->
